in-progress/8757-fees/design.md

```diff
@@ -0,0 +1,270 @@
+- `L2_SLOTS_PER_L2_EPOCH` - the number of L2 slots in each L2 epoch (e.g. 32)
+- `L1_GAS_PER_BLOCK_PROPOSED` - the amount of L1 gas required to propose an L2 block on L1 (e.g. 0.2e6)
+- `L1_GAS_PER_EPOCH_VERIFIED` - the amount of L1 gas required to verify an L2 epoch on L1 (e.g. 1e6)
+- `MAXIMUM_EPOCH_PROOF_QUOTE_FEE` - the maximum basis point fee a prover can submit for an epoch proof quote (e.g. 9000)
+- `UNDERLYING_BASE_FEE_ORACLE_UPDATE_INTERVAL` - the minimum number of slots between updates to the underlying base fee oracle (e.g. 4)
+- `MINIMUM_PROVING_COST_PER_MANA` - the minimum cost in wei for proving a unit of mana (e.g. 100)
+- `MAXIMUM_PROVING_COST_PERCENT_CHANGE_PER_BLOCK` - the maximum percentage increase in the cost of proving a unit of mana per block (e.g. 1%)
+- `PROVING_COST_UPDATE_FRACTION` - a value used to update the `proving_cost_modifier` (e.g. 1e11)
+- `MINIMUM_FEE_ASSET_PRICE` - the minimum price in wei of the fee asset (e.g. TST)
+- `MAXIMUM_FEE_ASSET_PRICE_PERCENT_CHANGE_PER_BLOCK` - the maximum percentage increase in the price of the fee asset per block (e.g. 1%)
+- `FEE_ASSET_PRICE_UPDATE_FRACTION` - a value used to update the `fee_asset_price_modifier` (e.g. 1e11)
+- `MINIMUM_CONGESTION_MULTIPLIER` - the minimum value the congestion multiplier can take (e.g. 1)
+- `CONGESTION_MULTIPLIER_UPDATE_FRACTION` - the constant factor to dampen movement in the congestion multiplier (e.g. 8.547 * TARGET_MANA_PER_BLOCK)
+
+## Block Header Fields
+The L2 block header contains the following fields:
+
+- `total_mana_used` - the total mana used by the block
+- `base_fee_asset_per_mana` - the base fee in fee asset per mana
+
+## Rollup Contract Fields
+
+The rollup contract contains the following fields:
+
+- `proving_cost_modifier` - a value used in the computation of the proving cost per mana
+- `proving_cost_modifier_delta` - an "oracle" used to update the `proving_cost_modifier` from the previous block
+- `fee_asset_price_modifier` - a value used in the computation of the fee asset price per wei
+- `fee_asset_price_modifier_delta` - an "oracle" used to update the `fee_asset_price_modifier` from the previous block
+- `wei_per_l1_gas` - the cost of L1 gas in wei. Updated by anyone to the current L1 gas price at most every `UNDERLYING_BASE_FEE_ORACLE_UPDATE_INTERVAL` slots
+- `wei_per_l1_blob_gas` - the cost of L1 blob gas in wei. Updated by anyone to the current L1 blob gas price at most every `UNDERLYING_BASE_FEE_ORACLE_UPDATE_INTERVAL` slots
+
+## L1 Execution Cost
+
+The cost of a block's execution in wei is equal to the following:
+```math
+\begin{aligned}
+\text{execution cost in gas} &= \text{L1 GAS PER BLOCK PROPOSED}  \\
+&+ \text{BLOBS PER BLOCK} * \text{POINT EVALUATION PRECOMPILE GAS} \\
+&+ \frac{\text{L1 GAS PER EPOCH VERIFIED}}{\text{L2 SLOTS PER L2 EPOCH}} \\
+\text{execution cost in wei} &= \text{execution cost in gas} * \text{wei per l1 gas}
+\end{aligned}
+```
```

Comment:
Blobs per block is fixed at a constant because to hit 10 TPS we likely need 3 blobs per L2 block.

Missed slots shouldn't change the math above. We divide by that constant to have each block pay "its share" of of verification.